docs/power_manager.md

```diff
@@ -0,0 +1,120 @@
+
+    cd $HOME/vSwarm/tools/test-client && go build ./test-client.go
+
+    kn service apply -f $HOME/vSwarm/benchmarks/auth/yamls/knative/kn-auth-python.yaml
+    kn service apply -f $HOME/vSwarm/benchmarks/aes/yamls/knative/kn-aes-python.yaml
+    kn service apply -f $HOME/vSwarm/benchmarks/sleeping/yamls/knative/kn-sleeping-go.yaml
+    kn service apply -f $HOME/vSwarm/benchmarks/spinning/yamls/knative/kn-spinning-go.yaml
+    ```
+
+### Experiment 1: Workload sensitivity 
+2 node cluster is needed.
+1. On master node, run the node setup script:
+    ```bash
+    ./scripts/power_manager/setup_power_manager.sh;
+    ```
+   Then run the experiment:
+    ```bash
+    go run $HOME/vhive/examples/power_manager/workload_sensitivity/main.go;
+    ```
+
+### Experiment 2: Internode scaling
+3 node cluster is needed. 3 scenarios are performed:
+- Scenario 1: All worker nodes have low frequency 
+- Scenario 2: All worker nodes have high frequency
+- Scenario 3: 1 worker node has high frequency, another with low frequency (need to manually tune like experiment 3 point 4&5 below)
+
+1. On master node, run the node setup script:
+    ```bash
+    ./scripts/power_manager/setup_power_manager.sh;
+    ```
+   Then run the experiment:
+    ```bash
+    go run $HOME/vhive/examples/power_manager/internode_scaling/main.go;
+    ```
+
+### Experiment 3: Class Assignment 
+3 node cluster is needed. We need to able to assign the workload to specific node.
+
+1. Thus on master node, we need to enable nodeSelector:
+```bash
+   kubectl patch configmap config-features -n knative-serving -p '{"data": {"kubernetes.podspec-nodeselector": "enabled"}}'
+```
+
+2. We need to modify these benchmark knative yaml file to add nodeSelector in vSwarm before deployment (Spinning and AES to worker-high class, Sleeping and Auth to worker-low class). For example: 
+````yaml
+apiVersion: serving.knative.dev/v1
+kind: Service
+metadata:
+ name: spinning-go
+ namespace: default
+spec:
+ template:
+   spec:
+     nodeSelector:
+       loader-nodetype: worker-high
+     containers:
+       - image: docker.io/vhiveease/relay:latest
+         ports:
+           - name: h2c
+             containerPort: 50000
+         args:
+           - --addr=0.0.0.0:50000
+           - --function-endpoint-url=0.0.0.0
+           - --function-endpoint-port=50051
+           - --function-name=aes-go
+       - image: docker.io/kt05docker/sleeping-go:latest
+         args:
+           - --addr=0.0.0.0:50051    - 
+```` 
+
+3. On master node, label the worker node 
+    ```bash
+    kubectl label node node-1.kt-cluster.ntu-cloud-pg0.utah.cloudlab.us loader-nodetype=worker-low
+    kubectl label node node-2.kt-cluster.ntu-cloud-pg0.utah.cloudlab.us loader-nodetype=worker-high 
+    ```
+    Run the node setup script:
+    ```bash
+    ./scripts/power_manager/setup_power_manager.sh;
+    ```
+4. On worker node 1, manually set all CPU frequency to 1.2GHz. ie run the below command for all CPU core:
```

Comment:
Why are we doing it manually instead of using power profiles?

Reply:
because of the shared pool issue that was discussed last time, it only allows 1 PowerProfile at one time. So both nodes must share the same power profile (hence same freq) unless we do it manually for this experiment (since we want 50/50: 1 high freq node, 1 low freq node)

Reply:
If that's true, the power manager is almost useless since it fails to manage the whole cluster from a single point. If you are confident about it, please provide a link to documentation that says so and include it in the docs for the power profile part.

But I think it is your misconfiguration. In the power workload, we can choose the power profile. So, if there can be only one, why would we make a choice? I think you need to register/rename or do something else to add a second profile.